### PR TITLE
EventScript for when a deployment dir is cleared

### PR DESCRIPTION
**Pull request by AffectedArc07**
:cl: Core
Added `DeploymentClenaup` EventScript for when a deployment directory on disk is about to be removed.
/:cl:

Why? I am making a system where the server rsc is put onto a ramdisk to speed up icon operations, since some of them can take >100ms!!
image

Icon ops are highly dependant on disk IO, and going through a VM + questionable storage threading can lead to issues, not to mention how the RSC writes can be somewhat random. 

The `CompileComplete` eventscript allows me to move the rsc to the ramdisk and make a symlink to the game dir, but I need a way to clear that up since deleting the symlink wouldnt delete the rsc from the ramdisk, causing issues. This EventScript would allow me to do that.

---

Sidenote, I dont know a good way to test this, so this is currently untested outside of a compile. If you have any good pointers please tell me.

### CODE REVIEW

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Reviewed this last week but forgot to submit. Also merge dev, should be fixed now

File: src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs [hunk 1]
```diff
@@ -358,6 +367,7 @@ public async Task CleanUnusedCompileJobs(CancellationToken cancellationToken)
 				try
 				{
 					++deleting;
+					await eventConsumer.HandleEvent(EventType.DeploymentCleanedUp, new List<string> { x }, cancellationToken);
```
[the remark this answers is not in the document]
Comment (Cyberboss) [reply src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs]: Can you refactor it a little please. There's a lambda call `HandleCleanup` in the `CleanJob` function where this also happens. 

- Move the meat of that function into a class function that this code also calls. Call it `DeleteCompileJobContent`. Have it do these three actions (in parallel, `Task.WhenAll`), keep the remaining code in `HandleCleanup`:
  - Delete the directory
  - Mark the GitHub deployment as inactive
  - Run this new event
- Rename `CleanJob` to `CleanRegisteredCompileJob`

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
Actually, run the event script to completion before deleting the directory. GitHub deployment inactivation can be parallel with both

**Comment by AffectedArc07** [comment]
image

image

File: src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs [hunk 2]
```diff
@@ -426,5 +423,18 @@ async Task HandleCleanup()
 					logger.LogTrace("Compile job {0} lock count now: {1}", job.Id, decremented);
 				}
 		}
+
+		/// <summary>
+		/// Handles cleaning the resources of a <see cref="CompileJob"/>.
+		/// </summary>
+		/// <param name="directory">The directory to cleanup.</param>
+		/// <param name="cancellationToken">The <see cref="CancellationToken"/> for this <see cref="Task"/>.</param>
+		/// <returns>The deletion task</returns>
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs]:
```suggestion
		/// <returns>The deletion <see cref="Task"/>.</returns>
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs (hunk 2 above)
Comment (AffectedArc07) [reply src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs]: I floated this idea but thought it was excessive

lmao

File: src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs [hunk 3]
```diff
@@ -414,5 +423,18 @@ async Task HandleCleanup()
...
+		/// <summary>
+		/// Handles cleaning the resources of a <see cref="CompileJob"/>.
+		/// </summary>
+		/// <param name="directory">The directory to cleanup.</param>
+		/// <param name="cancellationToken">The <see cref="CancellationToken"/> for this <see cref="Task"/>.</param>
+		/// <returns>The deletion <see cref="Task"/>.</returns>
+		async Task DeleteCompileJobContent(string directory, CancellationToken cancellationToken)
+		{
+			// Then call the cleanup event, waiting here first
+			await eventConsumer.HandleEvent(EventType.DeploymentCleanup, new List<string> { directory }, cancellationToken);
```
Comment (Cyberboss) [inline src/Tgstation.Server.Host/Components/Deployment/DmbFactory.cs:436]:
```suggestion
			await eventConsumer.HandleEvent(EventType.DeploymentCleanup, new List<string> { ioManager.ResolvePath(directory) }, cancellationToken);
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**